src/aioswitcher/device/tools.py

```diff
@@ -165,6 +166,34 @@ def convert_str_to_devicetype(device_type: str) -> DeviceType:
     return DeviceType.MINI
 
 
+def convert_token_to_packet(
+    device_type: DeviceType, token: Union[str, None] = None
+) -> str:
+    """Return token packet from token if is valid, otherwise empty string."""
+    is_token_needed = bool(device_type.token_needed)
+    is_token_not_empty = token is not None and str(token) != ""
+    if is_token_needed:
+        if is_token_not_empty:
+            try:
+                token_packet = tok.he(token)
+            except (KeyError, ValueError) as ve:
+                raise RuntimeError("convert token to packet was not successful") from ve
+            return str(token_packet)
+        else:
+            raise RuntimeError("a token is needed but missing")
+    return ""
+
+
+def is_token_valid(device_type: DeviceType, token: Union[str, None] = None) -> bool:
+    """Return true if token is used and valid."""
+    is_token_needed = bool(device_type.token_needed)
+    is_token_not_empty = token is not None and str(token) != ""
+    if is_token_needed:
+        if is_token_not_empty:
+            return True
+    return False
```

Comment:
Looks like this entire function can be replaced with a simple statement, why do we need it as a function?

```python
device_key.token_needed and bool(token)
```

Reply:
Because I am calling this function many times when using the API calls.
I think it's better to write this once and just call it instead of adding this statement over and over in the code

Reply:
I'm afraid I have to disagree. It's new and not being used, so I don't see any point in having it as a function. What do you mean by `over and over`? How many junctions in this code do you expect will require this? This should be a lot to justify; it's just a simple ~40-charachters evaluation statement.

Reply:
If you think you have enough usage for this function to justify it. Could you include it with the code that's going to be used? There's no point in doing this early.

Reply:
I think there is enough usage in the next PR with the full RUNNER S11 integration and the next devices in the plan.
Here are some examples of usage for the next PR:
https://github.com/TomerFi/aioswitcher/pull/747/files#diff-cec958238bf6f0279cfc946924d9d860586123564684aaa276671879bcee8195R173

https://github.com/TomerFi/aioswitcher/pull/747/files#diff-cec958238bf6f0279cfc946924d9d860586123564684aaa276671879bcee8195R192

https://github.com/TomerFi/aioswitcher/pull/747/files#diff-cec958238bf6f0279cfc946924d9d860586123564684aaa276671879bcee8195R777
 
I think it makes the code that uses it more cleaner..

Reply:
I won't review a closed PR—it's closed for a reason. Please include this in the PR where you're using it if you think it's needed, and we can continue the discussion there.

For the record, I would like to clarify your request. You want to replace **3** occurrences of:
```python
device_type.token_needed and bool(token)
```

With **3** occurrences of:
```python
is_token_valid(self._device_type, self._token)
```
And this utility function:
```python
def is_token_valid(device_type: DeviceType, token: Union[str, None] = None) -> bool:
    "Return true if token is used and valid."""
    return device_type.token_needed and bool(token)
```

And this testing function:
```python
@mark.parametrize("type, token, is_token_valid", [
    (DeviceType.RUNNER, "", False),
    (DeviceType.RUNNER, "zvVvd7JxtN7CgvkD1Psujw==", False),
    (DeviceType.RUNNER, None, False),
    (DeviceType.RUNNER_S11, "zvVvd7JxtN7CgvkD1Psujw==", True),
    (DeviceType.RUNNER_S11, "", False),
    (DeviceType.RUNNER_S11, None, False)
    ])
def test_is_token_valid_should_return_expected_bool(type, token, is_token_valid):
    assert_that(tools.is_token_valid(type, token)).is_equal_to(is_token_valid)
```

---

I am trying to understand how this is considered cleaner.
The following is very readable to Python developers. If you think a clarification is in order, add a comment. 

```python
device_type.token_needed and bool(token) # this validates the token
```

Reply:
I didn't want to add the usage of this because there are dependants on that and I didn't want the PR become bigger..
There are more than 3 usages I just mentioned those as an examples.

I think this is a better approach and also covert by tests so why not?

Reply:
> I won't review a closed PR—it's closed for a reason. Please include this in the PR where you're using it if you think it's needed, and we can continue the discussion there.

Reply:
> > I won't review a closed PR—it's closed for a reason. Please include this in the PR where you're using it if you think it's needed, and we can continue the discussion there.

I can't add the usage parts as I mentioned.. it will need to bring all Runner S11 stuff with that which makes this PR big.
I am trying to make this PR small as possible.. I am not sure how to progress..

Reply:
I don't understand. Please clarify: Why can't you delete `is_token_valid` and its testing function from this PR despite them not being used at all?

Reply:
> I don't understand. Please clarify: Why can't you delete is_token_valid and its testing function from this PR despite them not being used at all?

I am not going to argue with you.. I removed the function and its test.